### ISSUE TITLE
Broken README image

### ISSUE THREAD

**Issue by daveverwer** [SwiftPackageIndex/SwiftPackageIndex-Server#2790]
We do some wrangling of GitHub image URLs to make them work in SPI READMEs but there's obviously something going on here that's breaking this one:

![Screenshot 2023-12-15 at 12 59 37@2x](https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/assets/5180/4f8622f6-be54-495a-b4e9-b3ca2261ac32)

and on GitHub:

![Screenshot 2023-12-15 at 13 00 39@2x](https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/assets/5180/5b4367bc-2718-4cf4-b93e-581384b8d45b)

In the GitHub page, the URL is:

```
https://private-user-images.githubusercontent.com/60271542/286382420-cf47949d-e99b-42f8-bd74-b308252edf59.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI2NDU1ODYsIm5iZiI6MTcwMjY0NTI4NiwicGF0aCI6Ii82MDI3MTU0Mi8yODYzODI0MjAtY2Y0Nzk0OWQtZTk5Yi00MmY4LWJkNzQtYjMwODI1MmVkZjU5LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE1VDEzMDEyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ0ZTk2N2UyYjE0ZjY0OTNjNTlkNjgxZWI0ODQzZjY5MTk2ZmY2NDczNTYxNDVkZGQyZjVjMDQwZWYxMTA4MTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.CQn8CPtaCbrZtPYS1KpNCgnX506Xkv4V5J1hfqOWCb8
```

and in our page, the URL is:

```
https://private-user-images.githubusercontent.com/60271542/286382420-cf47949d-e99b-42f8-bd74-b308252edf59.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI2NDQxMDcsIm5iZiI6MTcwMjY0MzgwNywicGF0aCI6Ii82MDI3MTU0Mi8yODYzODI0MjAtY2Y0Nzk0OWQtZTk5Yi00MmY4LWJkNzQtYjMwODI1MmVkZjU5LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE1VDEyMzY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ0NzQ5YTZkMGM0YmJiOTE2NTJlM2YxOTljZDY0YjVjN2Q0MDE3NTdmYTcxMWRlNjRjNmYwMTM2YjQzYTYzZDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2O283Gh1A5PqRJYi65hoPOuoywkibbC3pBxTA_TN8m0
```

The `jwt` is wrong in the second one. I thought that we may have just been out of sync with our README cache, but the repository hasn't been been touched in 5 days. Odd.

**Comment by daveverwer** [comment]
Another one:

![Screenshot 2023-12-15 at 14 09 48@2x](https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/assets/5180/01c52d74-85f3-43f7-b2c4-c104af8a81ea)

**Comment by daveverwer** [comment]
Another one https://swiftpackageindex.com/naftaly/Threadcrumb

**Comment by daveverwer** [comment]
I have noticed this getting more common, and unfortunately, the solution is not trivial.

When we request the GitHub README file, we get different image URLs than they use in their rendering of the README. Our images have a different URL and have a JWT attached. Unfortunately, that JWT expires 5 minutes after the API request completes. We cache the HTML, and therefore will never successfully show those images.

The only way to fix this is by caching any image with a JWT attached. This doesn’t affect every image hosted by GitHub, but the ones uploaded in this way are all affected.

The technique for doing this should be reasonably straightforward:

1. Get the HTML as we currently do in `Github.fetchReadme`
2. Parse it with SwiftSoup, which we already use
3. Find any image elements that have a source URL on the domain `private-user-images.githubusercontent.com` combined with having a JWT query parameter.
4. Cache any image that meets that criteria to the readme bucket in the same directory as the readme file but with a UUID as the image filename.
5. Replace the image URL with SwiftSoup
6. Store the readme HTML as usual.

I did a quick test on how many images are in this form and we have 601 images across 144 repositories in this format, so I don’t think the storage or access costs will be a problem.
